WallsAndHoles/xmltool.cpp

```diff
@@ -0,0 +1,283 @@
+#include "xmltool.h"
+#include <QDebug>
+#include <QMessageBox>
+
+SharedTileMap XMLTool::openTileMap(QString tileMapPath){
+    //load the file
+    QFile file(tileMapPath);
+    if (!file.exists() || !file.open(QFile::ReadOnly | QFile::Text)) {
+        qDebug() << "Fail to open file:"+tileMapPath;
+        return nullptr;
+    }
+    QXmlStreamReader xmlReader(file.readAll());
+    SharedTileMap tileMap;
+    //Parse the XML until we reach end of it
+    while(!xmlReader.atEnd() && !xmlReader.hasError()) {
+        // Read next element
+        QXmlStreamReader::TokenType token = xmlReader.readNext();
+        //If token is just StartDocument - go to next
+        if(token == QXmlStreamReader::StartDocument) {
+                continue;
+        }
+        //If token is StartElement - read it
+        if(token == QXmlStreamReader::StartElement) {
+            if(xmlReader.name() == "TileMap") {
+                QSize size;
+                foreach(const QXmlStreamAttribute &attr, xmlReader.attributes()) {
+                    if (attr.name().toString() == QString("width")) {
+                         size.setWidth(attr.value().toInt());
+                    }
+                    if (attr.name().toString() == QString("height")) {
+                         size.setHeight(attr.value().toInt());
+                    }
+                }
+                tileMap = SharedTileMap::create(size);
+            }
+            if(xmlReader.name()=="TileTemplateSet"){
+                SharedTileTemplateSet templateSet;
+                foreach(const QXmlStreamAttribute &attr, xmlReader.attributes()) {
+                    if (attr.name().toString() == QString("SavePath")) {
+                        templateSet = XMLTool::openTileTemplateSet(attr.value().toString());
+                    }
+                }
+                if(templateSet==nullptr)
+                    return nullptr;
```

Comment:
For now this is okay, but VERY importantly there should be very few errors that cause the map to not be loaded at all. For this case, it could open a new dialog for the user to find the file of the TileTemplateSet that failed to open, or an option to remove any associated tiles.

This is something we can save for a future task though

Reply:
Since XMLTool is just simple function, I think it's better to allow it just return nullptr or -1 when failed, left all dialog code to editor.